pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -407,10 +402,9 @@ func (c *Controller) onNodeEvent(obj interface{}, event model.Event) error {
 	return nil
 }
 
-func isNodePortGatewayService(svc *model.Service) bool {
-	svc.Mutex.RLock()
-	defer svc.Mutex.RUnlock()
-	return svc.Attributes.ClusterExternalPorts != nil
+func isNodePortGatewayService(svc *v1.Service) bool {
+	_, ok := svc.Annotations[kube.NodeSelectorAnnotation]
+	return ok && svc.Spec.Type == v1.ServiceTypeNodePort
```

Comment:
Should we nuke this entire thing and simply collect node ips from the pods of the gateway?

Reply:
I have thought about when i wrote this pr and found that it is a little complicated.  I will work as a separate one.  This is more like an optimization